regularizer_new.py

```diff
@@ -103,83 +245,249 @@ def tokenizer_helper(s, add_special_tokens=True):
+                    fully_contextual_vectors = fully_contextual_vectors.squeeze()
+                elif (self.linear):
+                    fully_contextual_vectors = all_vector_idxs[st : en + 1].mean(
+                        axis=0
+                    )
+                else:
                     fully_contextual_vectors = all_vector_idxs[st : en + 1].sum(
                         axis=0
                     )
+                if (self.linear):
+                    if (st == 0):
+                        scores[key] = 0
+                        continue
+                    else:
+                        inner_context_vectors = all_vector_idxs[st]
+                else:
                     inner_context_vectors = inner_context_vecs[idx][-1]
-                    scores[str_idx][key] = self.sim_metric(
-                        inner_context_vectors, fully_contextual_vectors
-                    )
-        else:
-            all_vector_idxs = outer_context_vecs[0][-1]
-            for idx, key in slice_dict[0]:
-                st, en = key
-
-                fully_contextual_vectors = all_vector_idxs[st : en + 1].sum(
-                    axis=0
-                )
-                inner_context_vectors = inner_context_vecs[idx][-1]
                 scores[key] = self.sim_metric(
                     inner_context_vectors, fully_contextual_vectors
                 )
+        else:
+            for str_idx in range(str_num):
+                all_vector_idxs = outer_context_vecs[str_idx][-1]
+                for idx, key in slice_dict[str_idx]:
+                    st, en = key
 
-        return scores
-    
-    def get_score(self, score_chart, mean=True):
+                    if (self.diff):
+                        if (st == 0):
+                            fully_contextual_vectors = all_vector_idxs[en]
+                        else:
+                            fully_contextual_vectors = all_vector_idxs[en] - all_vector_idxs[st-1]
+                        fully_contextual_vectors = fully_contextual_vectors.squeeze()
+                    elif (self.linear):
+                        fully_contextual_vectors = all_vector_idxs[st : en + 1].mean(
+                            axis=0
+                        )
+                    else:
+                        fully_contextual_vectors = all_vector_idxs[st : en + 1].sum(
+                            axis=0
+                        )
+                    if (self.linear):
+                        if (st == 0):
+                            scores[str_idx][key] = 0
+                            continue
+                        else:
+                            inner_context_vectors = all_vector_idxs[st]
+                    else:
+                        inner_context_vectors = inner_context_vecs[idx][-1]
+                    scores[str_idx][key] = self.sim_metric(
+                        inner_context_vectors, fully_contextual_vectors
+                    )
```

Comment:
Dangerous to duplicate large sections like this. Large chunks like this can easily break in future commits / create unexpected errors.

Reply:
Please merge this with the section above!

Reply:
Moved into functions.